File: tests/integration/benchmark_climate4r.R
# SPDX-FileCopyrightText: 2026 European Centre for Medium-Range Weather Forecasts (ECMWF)
# SPDX-License-Identifier: Apache-2.0

# Benchmark climate4R against tests/integration/benchmark_climate4r.py.
#
# Required packages: loadeR and climate4R.climdex.
# Run from the repository root, after running the Python companion:
#
#   Rscript tests/integration/benchmark_climate4r.R

base_url <- "https://sites.ecmwf.int/repository/earthkit-climate"
datasets <- c(
  tasmax = "tasmax_ACCESS-CM2_ssp585_far_future.nc",
  tasmin = "tasmin_ACCESS-CM2_ssp585_far_future.nc",
  pr = "pr_ACCESS-CM2_ssp585_far_future.nc"
)
valid_indicators <- c("PRCPTOT", "SDII", "DTR")

default_data_dir <- function() {
  cache_root <- Sys.getenv("XDG_CACHE_HOME", unset = file.path(path.expand("~"), ".cache"))
  file.path(cache_root, "earthkit-climate", "climate4r-benchmark")
}

usage <- function() {
  cat(paste0(
    "Usage: Rscript tests/integration/benchmark_climate4r.R [options]\n\n",
    "Options:\n",
    "  --repeats N                 Timed runs per indicator (default: 5)\n",
    "  --indicators A,B,...        PRCPTOT, SDII, and/or DTR (default: all)\n",
    "  --data-dir PATH             Directory for shared input files\n",
    "  --output PATH               climate4R summary CSV path\n",
    "  --python-results PATH       Python summary CSV path\n",
    "  --comparison-output PATH    Combined comparison CSV path\n",
    "  --help                      Show this message\n"
  ))
}

parse_args <- function(arguments) {
  options <- list(
    repeats = 5L,
    indicators = valid_indicators,
    data_dir = default_data_dir(),
    output = "tests/integration/benchmark_results/climate4r.csv",
    python_results = "tests/integration/benchmark_results/earthkit_climate.csv",
    comparison_output = "tests/integration/benchmark_results/climate4r_comparison.csv"
  )
  names_by_flag <- c(
    "--repeats" = "repeats",
    "--indicators" = "indicators",
    "--data-dir" = "data_dir",
    "--output" = "output",
    "--python-results" = "python_results",
    "--comparison-output" = "comparison_output"
  )

  position <- 1L
  while (position <= length(arguments)) {
    flag <- arguments[[position]]
    if (flag == "--help") {
      usage()
      quit(status = 0L)
    }
    option_name <- unname(names_by_flag[flag])
    if (is.na(option_name) || position == length(arguments)) {
      stop("Unknown option or missing value: ", flag, call. = FALSE)
    }
    options[[option_name]] <- arguments[[position + 1L]]
    position <- position + 2L
  }

  options$repeats <- suppressWarnings(as.integer(options$repeats))
  if (is.na(options$repeats) || options$repeats < 1L) {
    stop("--repeats must be an integer of at least 1", call. = FALSE)
  }
  if (is.character(options$indicators) && length(options$indicators) == 1L) {
    options$indicators <- strsplit(options$indicators, ",", fixed = TRUE)[[1L]]
  }
  unknown <- setdiff(options$indicators, valid_indicators)
  if (length(unknown) > 0L) {
    stop("Unknown indicator(s): ", paste(unknown, collapse = ", "), call. = FALSE)
  }
  options
}

check_packages <- function() {
  required <- c("loadeR", "climate4R.climdex")
  missing <- required[!vapply(required, requireNamespace, logical(1L), quietly = TRUE)]
  if (length(missing) > 0L) {
    stop("Missing required R package(s): ", paste(missing, collapse = ", "), call. = FALSE)
  }
}

download_datasets <- function(data_dir) {
  dir.create(data_dir, recursive = TRUE, showWarnings = FALSE)
  paths <- file.path(data_dir, datasets)
  names(paths) <- names(datasets)
  for (variable in names(paths)) {
    destination <- paths[[variable]]
    if (file.exists(destination)) {
      next
    }
    temporary <- paste0(destination, ".part")
    cat("Downloading", basename(destination), "...\n")
    tryCatch(
      {
        utils::download.file(
          paste0(base_url, "/", datasets[[variable]]),
          destfile = temporary,
          mode = "wb",
          quiet = FALSE
        )
        if (!file.rename(temporary, destination)) {
          stop("Could not move downloaded file to ", destination)
        }
      },
      finally = if (file.exists(temporary)) unlink(temporary)
    )
  }
  paths
}

load_inputs <- function(paths) {
  list(
    tasmax = loadeR::loadGridData(dataset = paths[["tasmax"]], var = "tasmax", dictionary = FALSE),
    tasmin = loadeR::loadGridData(dataset = paths[["tasmin"]], var = "tasmin", dictionary = FALSE),
    pr = loadeR::loadGridData(dataset = paths[["pr"]], var = "pr", dictionary = FALSE)
  )
}

run_indicator <- function(indicator, inputs) {
  if (indicator == "PRCPTOT" || indicator == "SDII") {
    return(climate4R.climdex::climdexGrid(
      index.code = indicator,
      pr = inputs$pr,
      cal = "gregorian",
      parallel = FALSE
    ))
  }
  climate4R.climdex::climdexGrid(
    index.code = indicator,
    tn = inputs$tasmin,
    tx = inputs$tasmax,
    cal = "gregorian",
    parallel = FALSE
  )
}

validate_result <- function(result, indicator) {
  if (is.null(result$Data) || length(result$Data) == 0L || !any(is.finite(result$Data))) {
    stop(indicator, " returned no finite values", call. = FALSE)
  }
}

benchmark_indicator <- function(indicator, inputs, repeats) {
  cat("Warming up", indicator, "...\n")
  validate_result(run_indicator(indicator, inputs), indicator)

  timings <- numeric(repeats)
  for (repeat_number in seq_len(repeats)) {
    gc(verbose = FALSE)
    started <- proc.time()[["elapsed"]]
    result <- run_indicator(indicator, inputs)
    validate_result(result, indicator)
    timings[[repeat_number]] <- proc.time()[["elapsed"]] - started
    cat(sprintf("  %s %d/%d: %.6f s\n", indicator, repeat_number, repeats, timings[[repeat_number]]))
  }

  data.frame(
    indicator = indicator,
    library = "climate4R",
    mean_seconds = mean(timings),
    median_seconds = stats::median(timings),
    std_seconds = sqrt(mean((timings - mean(timings))^2)),
    repeats = repeats,
    stringsAsFactors = FALSE
  )
}

write_results <- function(results, output) {
  dir.create(dirname(output), recursive = TRUE, showWarnings = FALSE)
  utils::write.csv(results, output, row.names = FALSE)
  cat("Results written to", output, "\n")
}

write_comparison <- function(r_results, python_results_path, output) {
  if (!file.exists(python_results_path)) {
    cat("Python results not found at", python_results_path, "- skipping combined comparison.\n")
    return(invisible(NULL))
  }

  python_results <- utils::read.csv(python_results_path, stringsAsFactors = FALSE)
  python_results <- python_results[python_results$library == "earthkit-climate", ]
  comparison <- merge(
    python_results[c("indicator", "median_seconds")],
    r_results[c("indicator", "median_seconds")],
    by = "indicator",
    suffixes = c("_earthkit", "_climate4r")
  )
  comparison$earthkit_speedup <- comparison$median_seconds_climate4r / comparison$median_seconds_earthkit
  comparison <- comparison[order(match(comparison$indicator, valid_indicators)), ]

  dir.create(dirname(output), recursive = TRUE, showWarnings = FALSE)
  utils::write.csv(comparison, output, row.names = FALSE)
  cat("\nMedian-time comparison (speedup > 1 means earthkit-climate is faster):\n")
  print(comparison, row.names = FALSE)
  cat("Comparison written to", output, "\n")
}

main <- function() {
  options <- parse_args(commandArgs(trailingOnly = TRUE))
  check_packages()
  cat("R", R.version.string, "; repeats=", options$repeats, "\n", sep = "")
  paths <- download_datasets(options$data_dir)
  inputs <- load_inputs(paths)
  rows <- lapply(options$indicators, benchmark_indicator, inputs = inputs, repeats = options$repeats)
  results <- do.call(rbind, rows)
  write_results(results, options$output)
  write_comparison(results, options$python_results, options$comparison_output)
}

main()

File: tests/integration/benchmark_climate4r.py
# SPDX-FileCopyrightText: 2026 European Centre for Medium-Range Weather Forecasts (ECMWF)
# SPDX-License-Identifier: Apache-2.0

"""Benchmark earthkit-climate against the same workload used by climate4R.

Run this script from the repository root before running the R companion::

    python tests/integration/benchmark_climate4r.py
    Rscript tests/integration/benchmark_climate4r.R

Both scripts exclude downloading and data loading from the timed region. Their
default outputs are written under ``tests/integration/benchmark_results``.
"""

from __future__ import annotations

import argparse
import csv
import gc
import os
import statistics
import sys
import time
import urllib.request
import warnings
from collections.abc import Callable, Sequence
from pathlib import Path
from typing import Any

BASE_URL = "https://sites.ecmwf.int/repository/earthkit-climate"
DATASETS = {
    "tasmax": "tasmax_ACCESS-CM2_ssp585_far_future.nc",
    "tasmin": "tasmin_ACCESS-CM2_ssp585_far_future.nc",
    "pr": "pr_ACCESS-CM2_ssp585_far_future.nc",
}
INDICATORS = ("PRCPTOT", "SDII", "DTR")
RESULT_FIELDS = ("indicator", "library", "mean_seconds", "median_seconds", "std_seconds", "repeats")


def default_data_dir() -> Path:
    """Return the shared cache directory used by the Python and R scripts."""
    cache_root = Path(os.environ.get("XDG_CACHE_HOME", Path.home() / ".cache"))
    return cache_root / "earthkit-climate" / "climate4r-benchmark"


def download_datasets(data_dir: Path) -> dict[str, Path]:
    """Download missing sample files and return their local paths."""
    data_dir.mkdir(parents=True, exist_ok=True)
    paths: dict[str, Path] = {}
    for variable, filename in DATASETS.items():
        path = data_dir / filename
        paths[variable] = path
        if path.exists():
            continue

        temporary_path = path.with_suffix(f"{path.suffix}.part")
        print(f"Downloading {filename} ...")
        try:
            urllib.request.urlretrieve(f"{BASE_URL}/{filename}", temporary_path)
            temporary_path.replace(path)
        finally:
            temporary_path.unlink(missing_ok=True)
    return paths


def load_inputs(paths: dict[str, Path]) -> dict[str, Any]:
    """Load each input fully so disk access is outside the timed region."""
    import xarray as xr

    datasets: dict[str, Any] = {}
    for variable, path in paths.items():
        with xr.open_dataset(path) as dataset:
            datasets[variable] = dataset.load()
    return datasets


def indicator_functions(data: dict[str, Any]) -> dict[str, Callable[[], Any]]:
    """Build the three earthkit-climate calls matching climate4R's indices."""
    import xarray as xr

    import earthkit.climate as ekc

    temperature = xr.merge(
        [data["tasmax"][["tasmax"]], data["tasmin"][["tasmin"]]],
        compat="override",
    )
    precipitation = data["pr"][["pr"]]
    return {
        "PRCPTOT": lambda: ekc.indicators.wet_precip_accumulation(
            ds=precipitation,
            thresh="1 mm/day",
            freq="YS",
        ),
        "SDII": lambda: ekc.indicators.daily_pr_intensity(
            ds=precipitation,
            thresh="1 mm/day",
            freq="YS",
            op=">=",
        ),
        "DTR": lambda: ekc.indicators.daily_temperature_range(ds=temperature, freq="MS"),
    }


def materialize_and_validate(result: Any, indicator: str) -> None:
    """Materialize lazy output and reject empty or entirely non-finite results."""
    import numpy as np
    import xarray as xr

    if hasattr(result, "compute"):
        result = result.compute()
    if isinstance(result, xr.Dataset):
        values = result.to_array().values
    elif isinstance(result, xr.DataArray):
        values = result.values
    else:
        values = np.asarray(result)
    if values.size == 0 or not np.isfinite(values).any():
        raise RuntimeError(f"{indicator} returned no finite values")


def benchmark(indicator: str, function: Callable[[], Any], repeats: int) -> dict[str, str | int | float]:
    """Warm up once, then time and summarize repeated executions."""
    print(f"Warming up {indicator} ...")
    materialize_and_validate(function(), indicator)

    timings: list[float] = []
    for repeat in range(1, repeats + 1):
        gc.collect()
        started = time.perf_counter()
        result = function()
        materialize_and_validate(result, indicator)
        elapsed = time.perf_counter() - started
        timings.append(elapsed)
        print(f"  {indicator} {repeat}/{repeats}: {elapsed:.6f} s")

    return {
        "indicator": indicator,
        "library": "earthkit-climate",
        "mean_seconds": statistics.mean(timings),
        "median_seconds": statistics.median(timings),
        "std_seconds": statistics.pstdev(timings),
        "repeats": repeats,
    }


def write_results(rows: list[dict[str, str | int | float]], output: Path) -> None:
    """Write benchmark summaries using the schema shared with the R script."""
    output.parent.mkdir(parents=True, exist_ok=True)
    with output.open("w", newline="", encoding="utf-8") as stream:
        writer = csv.DictWriter(stream, fieldnames=RESULT_FIELDS)
        writer.writeheader()
        writer.writerows(rows)
    print(f"Results written to {output}")


def parse_args(argv: Sequence[str] | None = None) -> argparse.Namespace:
    """Parse command-line options."""
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--repeats", type=int, default=5, help="Timed runs per indicator (default: 5).")
    parser.add_argument(
        "--indicators",
        nargs="+",
        choices=INDICATORS,
        default=list(INDICATORS),
        help="Indicators to run (default: all).",
    )
    parser.add_argument("--data-dir", type=Path, default=default_data_dir(), help="Directory for shared input files.")
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("tests/integration/benchmark_results/earthkit_climate.csv"),
        help="Summary CSV path.",
    )
    args = parser.parse_args(argv)
    if args.repeats < 1:
        parser.error("--repeats must be at least 1")
    return args


def main(argv: Sequence[str] | None = None) -> int:
    """Run the earthkit-climate half of the comparison."""
    args = parse_args(argv)
    warnings.filterwarnings("ignore", message=r"Variable does not have a .* attribute\.")
    print(f"Python {sys.version.split()[0]}; repeats={args.repeats}")
    paths = download_datasets(args.data_dir)
    data = load_inputs(paths)
    functions = indicator_functions(data)
    rows = [benchmark(name, functions[name], args.repeats) for name in args.indicators]
    write_results(rows, args.output)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
